community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactions.java

```diff
@@ -143,20 +143,31 @@ public KernelTransaction newInstance( KernelTransaction.Type type, SecurityConte
     {
         assertCurrentThreadIsNotBlockingNewTransactions();
         SecurityContext frozenSecurityContext = securityContext.freeze();
-        newTransactionsLock.readLock().lock();
         try
         {
-            assertDatabaseIsRunning();
-            TransactionId lastCommittedTransaction = transactionIdStore.getLastCommittedTransaction();
-            KernelTransactionImplementation tx = localTxPool.acquire();
-            StatementLocks statementLocks = statementLocksFactory.newInstance();
-            tx.initialize( lastCommittedTransaction.transactionId(),
-                    lastCommittedTransaction.commitTimestamp(), statementLocks, type, frozenSecurityContext, timeout );
-            return tx;
+            while ( !newTransactionsLock.readLock().tryLock( 1, TimeUnit.SECONDS ) )
+            {
+                assertDatabaseIsRunning();
+            }
+            try
+            {
+                assertDatabaseIsRunning();
+                TransactionId lastCommittedTransaction = transactionIdStore.getLastCommittedTransaction();
+                KernelTransactionImplementation tx = localTxPool.acquire();
+                StatementLocks statementLocks = statementLocksFactory.newInstance();
+                tx.initialize( lastCommittedTransaction.transactionId(), lastCommittedTransaction.commitTimestamp(),
+                        statementLocks, type, frozenSecurityContext, timeout );
+                return tx;
+            }
+            finally
+            {
+                newTransactionsLock.readLock().unlock();
+            }
         }
-        finally
+        catch ( InterruptedException ie )
         {
-            newTransactionsLock.readLock().unlock();
+            Thread.interrupted();
+            throw new RuntimeException( ie );
```

Comment:
You could throw `org.neo4j.graphdb.TransactionFailureException` instead, since this `RuntimeException` will otherwise bubble straight out through things like `GraphDatabaseService.beginTx()` etc.